qutebrowser/browser/downloads.py

```diff
@@ -894,110 +894,203 @@ def _on_data_changed(self, idx, *, webengine):
             qtutils.ensure_valid(end_index)
         self.dataChanged.emit(start_index, end_index)
 
-    def _raise_no_download(self, count):
-        """Raise an exception that the download doesn't exist.
+    def _select_downloads(self, count, all_, indexset, criterion, errmsg):
+        """Return selected downloads specified by indexset that satisfy the
+        criterion.
 
         Args:
-            count: The index of the download
+            count: count of download to get the actions applied against.
+            all_: whether all downloads should get the actions applied against.
+            indexset: Specify the set of download indexes to delete. E.g.
+                '2,5,6-last,3', where 'last' is special keyword that denotes
+                that latest download index. If you specify numbers smaller than
+                1, or greater than 'last', then those numbers will be silently
+                ignored.
+            criterion: A function that returns True only when satisfied.
+            errmsg: The postfix of the error message that is shown alongside
+                    list of downloads that do not match the criterion.
         """
-        if not count:
-            raise cmdexc.CommandError("There's no download!")
-        raise cmdexc.CommandError("There's no download {}!".format(count))
+        # test if there are any downloads at all
+        if len(self) == 0:
+            raise cmdexc.CommandError(
+                    "No downloads!".format(indexset)
+                  )
+
+        # support --all/-a
+        if all_ == True:
+            indexset = '1-last'
+            message.warning(
+                "'--all/-a' will be deprecated. Kindly use '0-last' instead."
+            )
+
+        # support count
+        if indexset is None:
+            if count == 0:
+                count = len(self)
+            indexset = str(count)
+
+        # parse the indexset into number intervals list
+        try:
+            indexset_p = utils.parse_numsettxt_into_numints(indexset, len(self))
+        except ValueError:
+            raise cmdexc.CommandError(
+                    "Invalid index set '{}'!".format(indexset)
+                  )
+
+        # identify downloads indexes that fall in the indexset
+        ins = utils.which_nums_in_numints(
```

Comment:
Not sure I understand. Do you mean something like:

    indexset_p = IndexSet('0-3,8-last')
    ins = indexset_p.intersection(range(1,len(self)+1))

?

I guess that class `IndexSet` to be defined in `utils.py`?

Last question, got any better name suggestions than `IndexSet`?